magefiles/utils.go

```diff
@@ -82,3 +87,149 @@ func retry(f func() error, attempts int, delay time.Duration) error {
 	}
 	return fmt.Errorf("reached maximum number of attempts (%d). error: %+v", attempts, err)
 }
+
+func commandExists(cmd string) bool {
+	// check a command is available in the system.
+	_, err := exec.LookPath(cmd)
+	return err == nil
+}
+
+func initKCPController() (*kcp.SuiteController, error) {
+	kubeClient, err := kubeCl.NewK8SClient()
+	if err != nil {
+		return &kcp.SuiteController{}, fmt.Errorf("error creating client-go %v", err)
+	}
+
+	kcpController, err := kcp.NewSuiteController(kubeClient)
+	if err != nil {
+		return &kcp.SuiteController{}, err
+	}
+
+	return kcpController, nil
+}
+
+func redHatSSOAuthentication() error {
+	// Authenticate to Red Hat SSO using an offline token
+	// curl \
+	//  --silent \
+	//  --header "Accept: application/json" \
+	//  --header "Content-Type: application/x-www-form-urlencoded" \
+	//  --data-urlencode "grant_type=refresh_token" \
+	//  --data-urlencode "client_id=cloud-services" \
+	//  --data-urlencode "refresh_token=${OFFLINE_TOKEN}" \
+	//  "https://sso.redhat.com/auth/realms/redhat-external/protocol/openid-connect/token"
+
+	// Check offline_token is set: it's required for SSO
+	OFFLINE_TOKEN := os.Getenv("OFFLINE_TOKEN")
+	if OFFLINE_TOKEN == "" {
+		return fmt.Errorf("OFFLINE_TOKEN is required for Red Hat SSO")
+	}
+
+	// Get current home dir
+	HOMEDIR, err := os.UserHomeDir()
+	if err != nil {
+		return err
+	}
+
+	// Build the auth request and execute
+	data := url.Values{}
+	data.Set("grant_type", "refresh_token")
+	data.Set("client_id", "cloud-services")
+	data.Add("refresh_token", OFFLINE_TOKEN)
+	encodedData := data.Encode()
+
+	req, err := http.NewRequest("POST", "https://sso.redhat.com/auth/realms/redhat-external/protocol/openid-connect/token", strings.NewReader(encodedData))
+	if err != nil {
+		return fmt.Errorf("can't complete Red Hat SSO: %v", err)
+	}
+	req.Header.Set("Accept", "application/json")
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return fmt.Errorf("can't complete Red Hat SSO: %v", err)
+	}
+	defer resp.Body.Close()
+
+	// Get response data
+	var response_data map[string]interface{}
+	err = json.NewDecoder(resp.Body).Decode(&response_data)
+	if err != nil {
+		return err
+	}
+
+	// Create oidc-login cache folder and files to save tokens
+	oidc_dir := HOMEDIR + "/.kube/cache/oidc-login/"
+	if _, err := os.Stat(oidc_dir); os.IsNotExist(err) {
+		err = os.MkdirAll(oidc_dir, 0700)
+		if err != nil {
+			return err
+		}
+	}
+
+	oidc_filename := oidc_dir + "de0b44c30948a686e739661da92d5a6bf9c6b1fb85ce4c37589e089ba03d0ec6"
```

Comment:
where is this string "de0b44c30948a686e739661da92d5a6bf9c6b1fb85ce4c37589e089ba03d0ec6" actually coming from? should it be in a variable/const?

Reply:
I've ported [this](https://github.com/redhat-appstudio/e2e-tests/blob/88a6db42d91e6ae79b809e79c024f26c1ba95d73/scripts/install-appstudio-kcp.sh#L219) from the bash script. I think it's just a random string for the cache file with tokens.